docs/proposals/generalize-modules.md

```diff
@@ -0,0 +1,66 @@
+---
+title: "Generalize Modules Service to make it extensible"
+linkTitle: "Generalize Modules Service to make it extensible"
+weight: 1
+slug: generalize-modules
+---
+
+- Author: @annanay25
+- Reviewers: @jtlisi, @pstibrany
+- Date: April 2020
+- Status: Review
+
+## Overview
+
+Cortex uses modules to start and operate services with dependencies. Inter-service dependencies are specified in a map and passed to a module manager which ensures that they are started in the right order of dependencies. While this works really well, the implementation is tied in specifically to the Cortex struct and is not flexible for use with other projects like Loki, which also require similar forms of dependency management.
+
+We would like to extend modules in cortex to a generic dependency management framework, that can be used by any project with no ties to cortex.
+
+## Specific goals
+
+- Framework should allow for reusing cortex modules and allow us to:
+  - Add new modules
+  - Overwrite the implementation of a current module
+  - Manage dependencies
+- Framework should allow for building an application from scratch using the `modules` package, with no dependencies on Cortex. For ex: Remove code from Loki that was copied from `pkg/cortex/cortex.go`.
+
+
+## Proposed Design
+
+### Modules package
+
+To make the modules package extensible, we need to abstract away any Cortex specific details from the module manager. The proposed design is to -
+
+- Make a new component `Manager`, which is envisioned to be a central manager for all modules of the application. It stores modules & dependencies, and will be housed under a new package `pkg/util/modules`. `Manager` has the following methods for interaction:
+```
+   func (m *Manager) RegisterModule(name string, deps []string, initFn func() (Service, error))
+   func (m *Manager) AddDependency(fromModule string, toModule string) error
```

Comment:
Just looking at the function signature it's not clear to me if `fromModule` depends on `toModule` or viceversa. Variable naming wise, you may consider:

```
func (m *Manager) AddDependency(name string, dependsOn string) error
```

Reply:
Extending Marco's suggestion:

```
func (m *Manager) AddDependency(name string, dependsOn... string) error
```

(making `dependsOn` variadic)

Reply:
I agree, this is cleaner. Having a single way of adding dependencies as per your other comment also LGTM. :)